File: hello-dart/lib/src/cli_parser.dart
import 'dart:io';
import 'package:args/args.dart';
import 'package:intl/intl.dart';
import 'types.dart';
import 'logger.dart';

/// Command-line argument parser
class CliParser {
  static Config? parse(List<String> arguments) {
    final parser = ArgParser()
      ..addOption('input', help: 'Path to input audio file (required)')
      ..addOption('output', help: 'Path to output audio file (optional)')
      ..addOption('server',
          defaultsTo: 'ws://localhost:8080/audio', help: 'WebSocket server URI')
      ..addFlag('verbose', abbr: 'v', help: 'Enable verbose logging')
      ..addFlag('help',
          abbr: 'h', help: 'Display help message', negatable: false);

    try {
      final results = parser.parse(arguments);

      if (results['help'] as bool) {
        _printHelp(parser);
        return null;
      }

      final inputPath = results['input'] as String?;
      if (inputPath == null) {
        Logger.error('--input is required');
        _printHelp(parser);
        return null;
      }

      // Validate input file
      if (!File(inputPath).existsSync()) {
        Logger.error('Input file does not exist: $inputPath');
        return null;
      }

      // Generate default output path if not provided
      String outputPath;
      if (results['output'] != null) {
        outputPath = results['output'] as String;
      } else {
        final timestamp = DateFormat('yyyyMMdd-HHmmss').format(DateTime.now());
        // Use platform-independent path separator
        final filename = inputPath.split(Platform.pathSeparator).last;
        outputPath =
            'audio${Platform.pathSeparator}output${Platform.pathSeparator}output-$timestamp-$filename';
      }

      final verbose = results['verbose'] as bool;
      Logger.setVerbose(verbose);

      return Config(
        inputPath: inputPath,
        outputPath: outputPath,
        serverUri: results['server'] as String,
        verbose: verbose,
      );
    } catch (e) {
      Logger.error('Error parsing arguments: $e');
      _printHelp(parser);
      return null;
    }
  }

  static void _printHelp(ArgParser parser) {
    print('''
Audio Stream Client

Usage: audio_stream_client [OPTIONS]

Options:
${parser.usage}

Examples:
  audio_stream_client --input audio/input/test.mp3
  audio_stream_client --input audio/input/test.mp3 --output /tmp/output.mp3
  audio_stream_client --input audio/input/test.mp3 --server ws://192.168.1.100:8080/audio --verbose
''');
  }
}

File: hello-csharp/src/CliParser.cs
using System;
using System.IO;

namespace AudioFileTransfer;

/// <summary>
/// Command-line argument parser
/// </summary>
public static class CliParser
{
    public static Config ParseArgs(string[] args)
    {
        string? inputPath = null;
        string? outputPath = null;
        string serverUri = "ws://localhost:8080/audio";
        bool verbose = false;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "--input":
                    if (i + 1 < args.Length)
                    {
                        inputPath = args[++i];
                    }
                    break;
                case "--output":
                    if (i + 1 < args.Length)
                    {
                        outputPath = args[++i];
                    }
                    break;
                case "--server":
                    if (i + 1 < args.Length)
                    {
                        serverUri = args[++i];
                    }
                    break;
                case "--verbose":
                case "-v":
                    verbose = true;
                    break;
                case "--help":
                case "-h":
                    ShowHelp();
                    Environment.Exit(0);
                    break;
                default:
                    Console.Error.WriteLine($"Unknown argument: {args[i]}");
                    ShowHelp();
                    Environment.Exit(1);
                    break;
            }
        }

        if (string.IsNullOrEmpty(inputPath))
        {
            Console.Error.WriteLine("Error: --input argument is required");
            ShowHelp();
            Environment.Exit(1);
        }

        if (!File.Exists(inputPath))
        {
            Console.Error.WriteLine($"Error: Input file does not exist: {inputPath}");
            Environment.Exit(1);
        }

        if (string.IsNullOrEmpty(outputPath))
        {
            string timestamp = DateTime.Now.ToString("yyyyMMdd-HHmmss");
            string filename = Path.GetFileName(inputPath);
            outputPath = Path.Combine("audio", "output", $"output-{timestamp}-{filename}");
        }

        return new Config
        {
            InputPath = inputPath,
            OutputPath = outputPath,
            ServerUri = serverUri,
            Verbose = verbose
        };
    }

    private static void ShowHelp()
    {
        Console.WriteLine("Audio Stream Cache Client - C# Implementation");
        Console.WriteLine();
        Console.WriteLine("Usage: AudioFileTransfer [options]");
        Console.WriteLine();
        Console.WriteLine("Options:");
        Console.WriteLine("  --input <path>     Input audio file path (required)");
        Console.WriteLine("  --output <path>    Output audio file path (optional)");
        Console.WriteLine("  --server <uri>     WebSocket server URI (default: ws://localhost:8080/audio)");
        Console.WriteLine("  --verbose, -v      Enable verbose logging");
        Console.WriteLine("  --help, -h         Show this help message");
        Console.WriteLine();
        Console.WriteLine("Example:");
        Console.WriteLine("  AudioFileTransfer --input audio/input/test.mp3");
        Console.WriteLine("  AudioFileTransfer --input audio/input/test.mp3 --server ws://192.168.1.100:8080/audio");
        Console.WriteLine("  AudioFileTransfer --input audio/input/test.mp3 --output /tmp/output.mp3 --verbose");
    }
}
